woocommerce/woocommerce-gateway-stripe
# Add support for Stripe Location API

Fixes #2122

## Changes proposed in this Pull Request:

This PR adds support for the following API endpoints for dealing with [terminal locations](https://stripe.com/docs/api/terminal/locations). These endpoints will be used by mobile apps to assign card readers to Stripe locations.

The following endpoints are added:
* `GET /wc/v3/wc_stripe/terminal/locations`: Get all locations
* `POST /wc/v3/wc_stripe/terminal/locations`: Create a new location
* `GET /wc/v3/wc_stripe/terminal/locations/store`: Get the default store location
* `GET /wc/v3/wc_stripe/terminal/locations/<id>`: Get a location
* `POST /wc/v3/wc_stripe/terminal/locations/<id>`: Update a location
* `DELETE /wc/v3/wc_stripe/terminal/locations/<id>`: Delete a location

These endpoints are designed to mimic the behavior of the same endpoints on WCPay, so that the mobile app can easily switch between which payment gateway to use for in-person payments.

One difference between the two: in WCPay, transients are used to cache the location list to avoid making subsequent calls to Stripe for them. This works in part because, since WCPay merchants don't have their own Stripe dashboard, they can't add new locations to Stripe manually and get out of sync. With the Stripe extension, this isn't true, so I didn't implement the transients for now (though I'm open to it if making too many requests to Stripe becomes a problem.)

Known issue:
- In WCPay, the `GET /terminal/locations` endpoint uses Stripe's API to return just the first page of locations. Stripe's API includes a `has_more` property to indicate whether there is an additional page of locations, but since WCPay returns an array of locations, there's no place to include the `has_more` property. This is likely fine for now since the mobile app uses only a single default location, but I think we should eventually update this endpoint to handle pagination and have WCPay do the same for consistency.

## Testing instructions

For making API requests, I found it helpful to install the [Basic Auth](https://github.com/WP-API/Basic-Auth) plugin to my test site so that I could make requests of the form `curl --user username:password 'URL'` to test the endpoints. But other options (e.g. token authentication) are possible for testing these routes.

1. Activate the Stripe extension, add Stripe API credentials to settings.
2. Perform a request to `GET /wc/v3/wc_stripe/terminal/locations`, ensure that you get back an empty list since there are no locations yet.
3. Perform a request to `POST /wc/v3/wc_stripe/terminal/locations` with a JSON payload of `{"display_name": "Test Location 1", "address": {"line1": "1 Example St.", "city": "Example City", "country": "US", "state": "CA", "postal_code": 94111}}`.
4. Verify that you get back a response from Stripe that includes all of the fields listed under the `data` key for "Create Location" in pdfdoF-er-p2. The response will also include an `id` for the terminal location: take note of this for use in future API requests.
5. Visit the Locations page of your Stripe dashboard: https://dashboard.stripe.com/test/terminal/locations
6. Verify that you see the `Test Location 1` location on your dashboard.
7. Perform a request to `GET /wc/v3/wc_stripe/terminal/locations`, ensure that you get back a list with the single location you created in (3). Verify that the response from Stripe includes all of the fields listed under the `data` key for "Get Locations" in pdfdoF-er-p2.
8. Perform a request to `GET /wc/v3/wc_stripe/terminal/locations/<ID>` where `<ID>` is the location ID you took note of in (4). Verify that the response from Stripe includes all of the fields listed under the `data` key for "Get Location" in pdfdoF-er-p2.
9. Perform a request to `POST /wc/v3/wc_stripe/terminal/locations/<ID>` where `<ID>` is the location ID you took note of in (4). Use the JSON payload `{"display_name": "Test Location 2", "address": {"line1": "2 Example St.", "city": "Example City", "country": "US", "state": "CA", "postal_code": 94111}}`. Verify that the response from Stripe includes all of the fields listed under the `data` key for "Update Location" in pdfdoF-er-p2.
10. Go back to your Stripe dashboard. Verify that the location is now called "Test Location 2". Click on the location and verify that the address for the location is now "2 Example St."
11. Perform a request to `DELETE /wc/v3/wc_stripe/terminal/locations/<ID>` where `<ID>` is the location ID you took note of in (4). Verify that the response from Stripe includes all of the fields listed under the `data` key for "Delete Location" in pdfdoF-er-p2.
12. Go back to your Stripe dashboard. Verify that the location is no longer present.
13. Go to WooCommerce settings (`/wp-admin/admin.php?page=wc-settings`) and clear out any address text fields if present.
13. Perform a request to `GET /wc/v3/wc_stripe/terminal/locations/store`.  Verify that the response from Stripe matches the "Missing Store Address" response under "Get Default Location" in pdfdoF-er-p2. 
14. Go back to WooCommerce settings and add an example address.
15. Perform a request to `GET /wc/v3/wc_stripe/terminal/locations/store`.  Verify that the response from Stripe includes the fields listed under the `data` key for "Success" in "Get Default Location" in pdfdoF-er-p2. Verify that the response contains your store's name and the address you added in (14). Take note of the ID of the location given back to you in this step.
16. Check your Stripe dashboard and verify that your store's location is now present there.
17. Perform another request to `GET /wc/v3/wc_stripe/terminal/locations/store`. Verify that the response is the same as the response you got in (15), including the same ID.
18. Check your Stripe dashboard and verify that there's still just the one location there.
19. Change the address for your WooCommerce store.
20. Perform another request to `GET /wc/v3/wc_stripe/terminal/locations/store`. Verify that you now get back a different ID for your new store location.
21. Check your Stripe dashboard and verify that a second location has been added.

---


**Post merge**
